File: scripts/gen_icons_from_svg.py
"""Generate all Tauri icon sizes from logo.svg using svglib + Pillow."""
import os
import io
from svglib.svglib import svg2rlg
from reportlab.graphics import renderPM
from PIL import Image

ICONS_DIR = os.path.join(os.path.dirname(__file__), "..", "src-tauri", "icons")
SVG_PATH = os.path.join(ICONS_DIR, "logo.svg")


def svg_to_pil(svg_path, size=1024):
    """Convert SVG to a PIL Image at the given size."""
    drawing = svg2rlg(svg_path)
    # Scale to desired size
    sx = size / drawing.width
    sy = size / drawing.height
    drawing.width = size
    drawing.height = size
    drawing.scale(sx, sy)
    # Render to PNG bytes
    png_bytes = renderPM.drawToString(drawing, fmt="PNG", dpi=72)
    img = Image.open(io.BytesIO(png_bytes)).convert("RGBA")
    return img


def generate():
    print(f"Loading SVG from {SVG_PATH}")
    base = svg_to_pil(SVG_PATH, 1024)
    print(f"  Base image: {base.size}")

    # Required PNG sizes for Tauri
    png_icons = {
        "32x32.png": 32,
        "128x128.png": 128,
        "128x128@2x.png": 256,
        "icon.png": 512,
    }

    # Windows Store logos
    square_icons = {
        "Square30x30Logo.png": 30,
        "Square44x44Logo.png": 44,
        "Square71x71Logo.png": 71,
        "Square89x89Logo.png": 89,
        "Square107x107Logo.png": 107,
        "Square142x142Logo.png": 142,
        "Square150x150Logo.png": 150,
        "Square284x284Logo.png": 284,
        "Square310x310Logo.png": 310,
        "StoreLogo.png": 50,
    }

    all_pngs = {**png_icons, **square_icons}

    for filename, size in all_pngs.items():
        resized = base.resize((size, size), Image.LANCZOS)
        out_path = os.path.join(ICONS_DIR, filename)
        resized.save(out_path, "PNG")
        print(f"  Created {filename} ({size}x{size})")

    # ICO (Windows) - multiple sizes
    ico_sizes = [16, 24, 32, 48, 64, 128, 256]
    ico_images = [base.resize((s, s), Image.LANCZOS) for s in ico_sizes]
    ico_path = os.path.join(ICONS_DIR, "icon.ico")
    ico_images[0].save(
        ico_path,
        format="ICO",
        sizes=[(s, s) for s in ico_sizes],
        append_images=ico_images[1:],
    )
    print("  Created icon.ico")

    # ICNS (macOS)
    icns_path = os.path.join(ICONS_DIR, "icon.icns")
    icns_sizes = [16, 32, 64, 128, 256, 512, 1024]
    icns_images = [base.resize((s, s), Image.LANCZOS) for s in icns_sizes]
    try:
        icns_images[-1].save(
            icns_path, format="ICNS", append_images=icns_images[:-1]
        )
        print("  Created icon.icns")
    except Exception as e:
        print(f"  Warning: ICNS creation failed ({e}), saving as PNG fallback")
        base.resize((512, 512), Image.LANCZOS).save(icns_path, "PNG")
        print("  Created icon.icns (PNG fallback)")

    print("\nAll icons generated!")


if __name__ == "__main__":
    generate()

File: scripts/gen_icons_from_svg.mjs
/**
 * Generate all Tauri icon sizes from logo.svg using sharp.
 * Run: node scripts/gen_icons_from_svg.mjs
 */
import sharp from "sharp";
import { readFileSync, writeFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ICONS_DIR = join(__dirname, "..", "src-tauri", "icons");
const SVG_PATH = join(ICONS_DIR, "logo.svg");

const svgBuffer = readFileSync(SVG_PATH);

async function generatePng(size, filename) {
  await sharp(svgBuffer)
    .resize(size, size)
    .png()
    .toFile(join(ICONS_DIR, filename));
  console.log(`  Created ${filename} (${size}x${size})`);
}

async function generateIco(sizes, filename) {
  // ICO format: we'll use the png2ico approach via sharp
  // For simplicity, generate the largest PNG and use it
  // Tauri actually accepts a multi-size ICO or a single PNG
  const buffers = await Promise.all(
    sizes.map((s) => sharp(svgBuffer).resize(s, s).png().toBuffer())
  );

  // Build ICO file manually
  const icoHeader = Buffer.alloc(6);
  icoHeader.writeUInt16LE(0, 0); // Reserved
  icoHeader.writeUInt16LE(1, 2); // ICO type
  icoHeader.writeUInt16LE(buffers.length, 4); // Number of images

  const dirEntries = [];
  let dataOffset = 6 + buffers.length * 16;

  for (let i = 0; i < buffers.length; i++) {
    const s = sizes[i];
    const entry = Buffer.alloc(16);
    entry.writeUInt8(s >= 256 ? 0 : s, 0); // Width (0 = 256)
    entry.writeUInt8(s >= 256 ? 0 : s, 1); // Height
    entry.writeUInt8(0, 2); // Color palette
    entry.writeUInt8(0, 3); // Reserved
    entry.writeUInt16LE(1, 4); // Color planes
    entry.writeUInt16LE(32, 6); // Bits per pixel
    entry.writeUInt32LE(buffers[i].length, 8); // Size of image data
    entry.writeUInt32LE(dataOffset, 12); // Offset to image data
    dirEntries.push(entry);
    dataOffset += buffers[i].length;
  }

  const ico = Buffer.concat([icoHeader, ...dirEntries, ...buffers]);
  writeFileSync(join(ICONS_DIR, filename), ico);
  console.log(`  Created ${filename}`);
}

async function main() {
  console.log("Generating icons from logo.svg...\n");

  // Tauri required PNGs
  const pngIcons = [
    [32, "32x32.png"],
    [128, "128x128.png"],
    [256, "128x128@2x.png"],
    [512, "icon.png"],
  ];

  // Windows Store logos
  const squareIcons = [
    [30, "Square30x30Logo.png"],
    [44, "Square44x44Logo.png"],
    [71, "Square71x71Logo.png"],
    [89, "Square89x89Logo.png"],
    [107, "Square107x107Logo.png"],
    [142, "Square142x142Logo.png"],
    [150, "Square150x150Logo.png"],
    [284, "Square284x284Logo.png"],
    [310, "Square310x310Logo.png"],
    [50, "StoreLogo.png"],
  ];

  // Generate all PNGs
  for (const [size, name] of [...pngIcons, ...squareIcons]) {
    await generatePng(size, name);
  }

  // Generate ICO
  await generateIco([16, 24, 32, 48, 64, 128, 256], "icon.ico");

  // Generate ICNS - Pillow/macOS only, save a large PNG as fallback
  // Tauri will use it on macOS builds
  await sharp(svgBuffer)
    .resize(1024, 1024)
    .png()
    .toFile(join(ICONS_DIR, "icon.icns"));
  console.log("  Created icon.icns (PNG fallback for macOS)");

  console.log("\nAll icons generated!");
}

main().catch(console.error);
